vm: move the compiler interface into a separate module

zerbina opened the pull request:
## Summary
Move the parts that make up the direct interface between the compiler and the VM into a separate module, in order to reduce the amount of dependencies the `vm` module has on the rest of the compiler.

This is also a step towards completely separating the VM from `vmgen`.

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

@saem If this PR interferes with #484 too much, I can also turn it into a draft PR and wait with merging until #484 is finished.

## Notes for Reviewers
* I'm not entirely sure of the name `compilerbridge`

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented:
Thanks, @zerbina. #484 can wait, it'll have to get updated a bunch because of the report splits anyhow.

saem commented on compiler/sem/semstmts.nim:
```diff
@@ -915,7 +915,7 @@ proc semNormalizedLetOrVar(c: PContext, n: PNode, symkind: TSymKind): PNode =
       var x = newNodeI(result.kind, v.info)
       x.add producedDecl
-      vm.setupCompileTimeVar(c.module, c.idgen, c.graph, x)
+      setupCompileTimeVar(c.module, c.idgen, c.graph, x)
```
It's so obvious when I read it now. The base VM shouldn't have been doing this all along and the code was telling us.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    errorhandling,
+    errorreporting,
+    lineinfos,
+    reports,
```
Could you narrow this reports import with a `from` as well?

zerbina replied:
Yep, will do. That `report` import is a leftover from before rebasing.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+  ## `dest`. Implicit conversion is also performed, if necessary.
+  let t = formal.skipTypes(abstractInst+{tyStatic}-{tyTypeDesc})
+
+  # XXX: instead of performing conversion here manually, sem could generate a
```
This is a good little project for someone to dive into and learn.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    assert n.kind == nkNilLit
+  of tyOpenArray:
+    # Handle `openArray` parameters the same way they're handled elsewhere
+    # in the VM: simply pass the argument without an conversion
```
```suggestion
    # in the VM: simply pass the argument without a conversion
```

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    result = c.graph.emptyNode
+
+proc evalExpr*(c: var TCtx, n: PNode): PNode =
+  # deadcode
```
Mark with a deadcode pragma like debug/deprecated?

zerbina replied:
I removed it instead. It hasn't been in use for some time now, and I don't think this will change in the near future.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+
+  result = execute(c[], start, tos, cb).unpackResult(c.config, n)
+
+  c.mode = oldMode
```
Do we trust defer?

zerbina replied:
In the sense that it should be used here? I personally wouldn't do so, but am also not strongly against it.

zerbina replied:
Oh, I now see what you mean. The procedure can return early, in which case the previous `mode` wouldn't be restored. Yep, `defer` definitely makes sense then.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+
+  c.mode = oldMode
+
+proc evalConstExpr*(module: PSym; idgen: IdGenerator; g: ModuleGraph; e: PNode): PNode =
```
Inline?

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    n.typ = x.typ
+    reg = TFullReg(kind: rkNimNode, nimNode: n)
+
+# XXX: why is this routine located here? It not related to the VM at all
```
Could put in AST query?

Aside: got this far in my code review

zerbina replied:
Done. The comment in question was new, but the rest of the code is for the largest part just copied over from `vm`.

saem reviewed (approved):
Minor changes, didn't get to review it all, but I'm already a big fan.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+                    n: PNode, sym: PSym): PNode =
+  #if g.config.errorCounter > 0: return errorNode(idgen, module, n)
+
+  # XXX globalReport() is ugly here, but I don't know a better solution for now
```
The global errors could be pushed out as nkError, then have the can raise the global there, keep applying this process inductively (? we're going up so not recursion, right? 😄) and we end up handling/recovering at the right spot.

Also applies to the next global error me thinks.

To be super clear, actual work of making that change doesn't need to be in this PR, just capturing the idea assuming it makes sense.

zerbina replied:
Actually, the `globalReport` for nested expansion is never reached -- `evalMacroCall` is not recursive. The only way for macros to expand other macros is via `getAst`, but that uses implemented via normal in-VM procedure calls, which already use their own mechanism for preventing recursion (albeit with a much higher limit).

I'm going to make a follow up PR that removes `evalMacroCounter` and the associated report.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    else:
+      # TODO: the decrement here is wrong, but the else branch is likely
+      #       currently not reached anyway
+      dec(g.config.evalMacroCounter)
```
If we trust `defer` then we're all good and can drop this line and the one at the end?

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,476 @@
+    setupMacroParam(tos.slots[i], c[], n[i], sym.typ[i])
+
+  # put macro generic parameters into registers
+  let gp = sym.ast[genericParamsPos]
```
Throwaway remark: I no longer recall the mechanics but it's weird that we put the generic params after the regular.

zerbina replied:
That's mainly a `vmgen` limitation. Normal parameters are always expected in register `param.position + 1`, so for macros, the generic parameters have to be placed after the normal ones.

saem commented on compiler/vm/compilerbridge.nim:
```diff
@@ -0,0 +1,458 @@
+    localReport(c.config, sym.info):
+      VMReport(kind: rvmCallingNonRoutine, sym: sym)
+
+# XXX: the compilerapi regarding globals (getGlobalValue/setGlobalValue)
```
Good

saem reviewed (approved):
Last pass done.

zerbina committed:
vm: move the compiler interface into a separate module

## Summary
Move the parts that make up the direct interface between the compiler
and the VM into a separate module, in order to reduce the amount of
dependencies the `vm` module has on the rest of the compiler.

This is also a step towards completely separating the VM from
`vmgen`.

## Details
- move the `genericParamsInMacroCall` iterator into the `ast_query`
  module
- remove the unused `evalExpr` procedure
- restore the previous `mode` value in `evalConstExprAux`
- restore the previous `mode` and `callsite` values in `evalMacroCall`

zerbina commented:
bors r+
